tests/src/App/IndexedName.cpp

```diff
@@ -0,0 +1,248 @@
+    auto indexedName = Data::IndexedName("TestName", testIndex);
+    EXPECT_STREQ(indexedName.getType(), "TestName");
+    EXPECT_EQ(indexedName.getIndex(), testIndex);
+}
+
+// Names must only contain ASCII letters and underscores (but may end with a number)
+TEST_F(IndexedNameTest, constructionInvalidCharInName)
+{
+    constexpr int lastASCIICode{127};
+    std::vector<char> illegalCharacters = {};
+    for (int code = 1; code <= lastASCIICode; ++code) {
+        if ((std::isalnum(code) == 0) && code != '_') {
+            illegalCharacters.push_back(char(code));
+        }
+    }
+    for (auto illegalChar : illegalCharacters) {
+        std::string testName {"TestName"};
+        testName += illegalChar;
+        auto indexedName = Data::IndexedName(testName.c_str(), 1);
+        EXPECT_STREQ(indexedName.getType(), "") << "Expected empty name when given " << testName;
+    }
+}
+
+// Names must not contain numbers in the middle:
+TEST_F(IndexedNameTest, constructionNumberInName)
+{
+    const int testIndex {42};
+    std::string testName;
+    testName += "Test" + std::to_string(testIndex) + "Name";
+    auto indexedName = Data::IndexedName(testName.c_str(), testIndex);
+    EXPECT_STREQ(indexedName.getType(), "");
+}
+
+TEST_F(IndexedNameTest, nameAndTypeListConstructionWithoutAllowOthers)
+{
+    auto indexedName = Data::IndexedName("EDGE19", allowedTypes, false);
+    EXPECT_STREQ(indexedName.getType(), "EDGE");
+    EXPECT_EQ(indexedName.getIndex(), 19);
+    indexedName = Data::IndexedName("EDGES_ARE_REALLY_GREAT19", allowedTypes, false);
+    EXPECT_STREQ(indexedName.getType(), "");
+    EXPECT_EQ(indexedName.getIndex(), 19);
+    indexedName = Data::IndexedName("NOT_IN_THE_LIST42", allowedTypes, false);
+    EXPECT_STREQ(indexedName.getType(), "");
+}
+
+TEST_F(IndexedNameTest, nameAndTypeListConstructionWithAllowOthers)
+{
+    auto indexedName = Data::IndexedName("NOT_IN_THE_LIST42", allowedTypes, true);
+    EXPECT_STREQ(indexedName.getType(), "NOT_IN_THE_LIST");
+    EXPECT_EQ(indexedName.getIndex(), 42);
+}
+
+TEST_F(IndexedNameTest, byteArrayConstruction)
+{
+    QByteArray qba{"EDGE42"};
+    auto indexedName = Data::IndexedName(qba);
+    EXPECT_STREQ(indexedName.getType(), "EDGE");
+    EXPECT_EQ(indexedName.getIndex(), 42);
+}
+
+TEST_F(IndexedNameTest, copyConstruction)
+{
+    auto indexedName = Data::IndexedName("EDGE42");
+    auto indexedNameCopy {indexedName};
+    EXPECT_EQ(indexedName, indexedNameCopy);
+}
+
+TEST_F(IndexedNameTest, streamInsertionOperator)
+{
+    auto indexedName = Data::IndexedName("EDGE42");
+    std::stringstream ss;
+    ss << indexedName;
+    EXPECT_EQ(ss.str(), std::string{"EDGE42"});
+}
+
+TEST_F(IndexedNameTest, compoundAssignmentOperator)
+{
+    constexpr int base{42};
+    constexpr int offset{10};
+    // Only += is defined for this class
+    auto indexedName = Data::IndexedName("EDGE",base);
+    indexedName += offset;
+    EXPECT_EQ(indexedName.getIndex(), base+offset);
```

Comment:
This will come off a being picky, sorry about that. In assertions it is good practice to avoid operations _just in case_ there is some non-obvious operator overloading somewhere.  So, it might better as `EXPECT_EQ(indexedName.getIndex(), 52);` . But that is paranoia speaking, feel free to leave it as is :).

Reply:
Pickiness is welcome here, these couple of classes are the core of the whole thing - IMO it's worth the time to get them right. I wrote it like that mostly to shut up the linter (which complains about the magic number otherwise), but I can just as easily add a NOLINT directive. Really, tests are littered with magic numbers, I am inclined to add that ignore to the entire file.